tutorials/tour/_posts/2017-02-13-pattern-matching.md

```diff
@@ -10,38 +10,129 @@ num: 12
-    case 1 => "one"
-    case "two" => 2
-    case y: Int => "scala.Int"
+abstract class Notification
+
+case class Email(sender: String, title: String, body: String) extends Notification
+
+case class SMS(caller: String, message: String) extends Notification
+
+case class VoiceRecording(contactName: String, link: String) extends Notification
+
+
+```
+`Notification` is an abstract super class which has three concrete Notification types implemented with case classes `Email`, `SMS`, and `VoiceRecording`. Now we can do pattern matching on these case classes:
+
+```
+def showNotification(notification: Notification): String = {
+  notification match {
+    case Email(email, title, _) =>
+      s"You got an email from $email with title: $title"
+    case SMS(number, message) =>
+      s"You got an SMS from $number! Message: $message"
+    case VoiceRecording(name, link) =>
+      s"you received a Voice Recording from $name! Click the link to hear it: $link"
+  }
+}
+val someSms = SMS("12345", "Are you there?")
+val someVoiceRecording = VoiceRecording("Tom", "voicerecording.org/id/123")
+
+println(showNotification(someSms))  // prints You got an SMS from 12345! Message: Are you there?
+
+println(showNotification(someVoiceRecording))  // you received a Voice Recording from Tom! Click the link to hear it: voicerecording.org/id/123
+```
+The function `showNotification` takes as a parameter the abstract type `Notification` and matches on the type of `Notification` (i.e. it figures out whether it's an `Email`, `SMS`, or `VoiceRecording`). In the `case Email(email, title, _)` the fields `email` and `title` are used in the return value but the `body` field is ignored with `_`.
+
+## Pattern guards
+```
+def showNotificationSpecial(notification: Notification, specialEmail: String, specialNumber: String): String = {
+  notification match {
+    case Email(email, _, _) if email == specialEmail =>
+      "You got an email from special someone!"
+    case SMS(number, _) if number == specialNumber =>
+      "You got an SMS from special someone!"
+    case other =>
+      showNotification(other) // nothing special, delegate to our original showNotification function   
   }
-  println(matchTest("two"))
+}
+
+val SPECIAL_NUMBER = "55555"
+val SPECIAL_EMAIL = "jane@mail.com"
+val someSms = SMS("12345", "Are you there?")
+val someVoiceRecording = VoiceRecording("Tom", "voicerecording.org/id/123")
+val specialEmail = Email("jane@mail.com", "Drinks tonight?", "I'm free after 5!")
+val specialSms = SMS("55555", "I'm here! Where are you?")
+
+println(showNotificationSpecial(someSms, SPECIAL_EMAIL, SPECIAL_NUMBER))
+println(showNotificationSpecial(someVoiceRecording, SPECIAL_EMAIL, SPECIAL_NUMBER))
+println(showNotificationSpecial(specialEmail, SPECIAL_EMAIL, SPECIAL_NUMBER))
+println(showNotificationSpecial(specialSms, SPECIAL_EMAIL, SPECIAL_NUMBER))
+```
+
+## Matching on type
+It is also possible to match on type.
+```tut
+def printType(x: Any): Unit = x match {
```

Comment:
ten minutes til @tpolecat shows up and asks if we have to use an example based on `Any`.

I'm not really sure what the right way to handle this is. The feature exists, it definitely should be included in this section of the tour, and we don't want to say it's always bad, but we don't want to give the impression that matching on `Any` is normal, either. perhaps up above, before showing case class extraction, you could show something like:

```
notification match {
    case e: Email => "you've got mail: $e"
    case s: SMS => "someone texted you: $s"
    case v: VoiceRecording => "check your voicemail: $v"
}
```

and then quickly move on and say "but most of the time, you don't just want to know what type something is, you want to do something with the individual fields, too", and then show the unapply-based version?

leaving the existing progression alone but just coming up with a non-`Any` based example also seems like a possibility?